crypto/dh/dh_lib.c

```diff
@@ -59,14 +59,15 @@ DH *DH_new_method(ENGINE *engine)
     ret->meth = DH_get_default_method();
 #ifndef OPENSSL_NO_ENGINE
     ret->flags = ret->meth->flags;  /* early default init */
-    if (engine) {
+    if (engine && ENGINE_get_DH(engine)) {
```

Comment:
I'm not familiar with the internal details of the engine framework but nevertheless this change looks conceptually wrong to me for two reasons:

- A call into the engine (`ENGINE_get_DH()`) is made before calling `ENGINE_init()` 
- This check competes with the assignment and check of `ENGINE_get_DH()` in lines 72,73.

If there is really a need to fallback to a default DH engine (which I can't decide from my current knowledge of the internals), then lines 68ff need to be refactored, too.